### DISCUSSION TITLE
Storybook MDX file does not display source

### DISCUSSION THREAD

**Discussion by konsalex** [storybookjs/storybook#24848] [Help] [answered]
Hey folks,

I am trying to write a Story with MDX format and for some reason I get `No code available`

![CleanShot 2023-11-15 at 11 06 13@2x](https://github.com/storybookjs/storybook/assets/12672541/bd54d1c5-6cbd-44af-b1cb-fa3984400041)

Any clue why this is happening.

The story looks like this:

```
import { useCallback, useRef, useState } from 'react';
import { Canvas, Meta, Story } from '@storybook/addon-docs';
import { Example } from './story';
import { useRenderCount } from '@neo4j-ndl/react/src/_common/hooks';
import { Button, TextInput } from '@neo4j-ndl/react/src';

<Meta title="Hooks/useRenderCount" />

# useRenderCount

This is a test

<Canvas>
  <Story name="Overview">
    {() => {
      const renderCount = useRenderCount();
      const [state, setState] = useState(0);
      return (
        <div className="n-flex n-flex-col n-gap-2 n-max-w-max">
          <Button onClick={() => setState(state + 1)}>Update state</Button>
          <TextInput value={renderCount} />
        </div>
      );
    }}
  </Story>
</Canvas>
``` 

And I followed the way Monday.com writes their MDX stories which seems to work as expected: https://github.com/mondaycom/monday-ui-react-core/blob/master/src/hooks/useClickableProps/__stories__/useClickableProps.stories.mdx

**Comment by ShaunEvening** [comment]
Hi @konsalex 👋 

Writing stories in MDX has been deprecated in Storybook 7. Do you have the same issue if you write your story in the CSF format instead?

**Reply by konsalex** [reply]
Technically it's not a story @Integrayshaun , rather than a code example 🤔  Would there be a way to achieve the same result? 

Basically writing a `<Code />` block that will show the content of the actual code, to demonstrate how to use the hook utilities. Could this be achieved somehow with CSF?

**Reply by ShaunEvening** [reply]
Do you need the example to run the code? If not, you could just put this code snippet into a regular markdown code block without needing the `<Canvas />` and `<Story />` components

**Reply by konsalex** [reply]
Yes would need to run the code, as it demonstrates the hook's usage in a visual way.

**Reply by ShaunEvening** [reply]
No worries we can do that!

You can write a CSF file for stories that aren't for a component 😁 Then you can import those stories and render them in your MDX docs.

Here's a StackBlitz example that I put together for you
https://stackblitz.com/edit/react-hook-story?file=src%2Fhooks%2FuseRenderCount%2FuseRenderCount.mdx

**Reply by konsalex** [reply]
@Integrayshaun thanks for the taking the time 🙏 Really appreciate it.

The example still not working, as you can see the `Show Code` tab is still a `{}` 🤔 

![CleanShot 2023-11-24 at 09 28 06@2x](https://github.com/storybookjs/storybook/assets/12672541/93265ea3-992b-4b62-9390-10473b1de130)

**Reply by ShaunEvening** [reply]
Right! Sorry about that.

So in this case if you want to show the code from the render function you will need to add it manually. When you put the render method into a component you'll get the usage of the component without any props like so:

<img width="1784" alt="image" src="https://github.com/storybookjs/storybook/assets/18172605/35283199-8f25-4d61-948b-79e5d2a19c94">

To instead show a code snippet that demonstrate the use of your hook, you can pass the code to the Canvas' `source` prop. I have updated the StackBlitz to show you how.

**Reply by konsalex** [reply, marked as answer]
@Integrayshaun thanks for the guidance. Ended up creating a pattern which is more maintainable as the code will live only in one place, and not manually pasted as a string.

```
story/
├── StoryBlock // The actual code that will run the hook example
├── hook.stories.tsx // The CSF compatible file
└── hook.mdx // The MDX doc of the hook, utilising Vite's `?raw` import flag
```

![CleanShot 2023-11-27 at 14 08 12@2x](https://github.com/storybookjs/storybook/assets/12672541/526d65bc-d9e3-49b4-b374-bf0325f36cd2)

Example for future reference from folks: 
https://stackblitz.com/edit/react-hook-story-cigb35?file=src%2Fhooks%2FuseRenderCount%2FStoryCode.tsx

**Reply by joevaugh4n** [reply]
Brilliant, great work @konsalex (and thanks for sharing your solution too)!

**Reply by konsalex** [reply]
Thanks @joevaugh4n . 

Is there any way with current Storybook's API to remove the story for the Sidebar, so only the Docs will be rendered? 🤔 

In our case have only the `useRenderCount` doc page and not the two stories alongside it:
![CleanShot 2023-11-27 at 17 18 52@2x](https://github.com/storybookjs/storybook/assets/12672541/78ba2c9b-4163-4cfe-9a35-d9f381b6cc96)

**Reply by konsalex** [reply]
Hacked around a bit more:
https://stackblitz.com/edit/react-hook-story-cigb35?file=.storybook%2Fmain.ts

That's the solution I found to remove the extra story, a bit hacky maybe?

So I use the `.mdx` file as a template, and then I export a single story called `Docs` which is then skipped due to name conflict and the proper doc is rendered. 

```tsx
import mdx from "./useRenderCount.mdx";

const meta = {
  title: 'Hooks/useRenderCount',
  render: StoryBlock,
  parameters: {
    docs: {
      page: mdx,
    },
  },
} satisfies Meta;

export default meta;

export const Docs = {}
```

But seems to work. 

![CleanShot 2023-11-28 at 10 08 20@2x](https://github.com/storybookjs/storybook/assets/12672541/117d435c-68d7-4ed7-9d02-e41cf5810e9e)

If you know some any better way please enlighten me

**Reply by dpavloff** [reply]
Hey @konsalex!

For your interest, passing `config.docs.docsMode: true` in main.js config will tell storybook to render only docs pages (either `.mdx` or `stories.` files)

```ts
const config: StorybookConfig = {
  ...
  
  docs: {
      autodocs: true,
      docsMode: true
    }
}
```

**Reply by roypearce** [reply]
@konsalex thank you for figuring this out, it's been a struggle to get things in newer versions of storybook to render in ways which are focused on just documentation for the stories which are purely for documentation and don't have components.

@dpavloff adding `docsMode` to the main config is an interesting option, but it applies globally, and at least for my project, we need stories for components to operate differently than stories for documentation that do not include components. Please clarify if I'm missing something.
